mockgen/model/model.go

```diff
@@ -147,12 +147,12 @@ type Type interface {
 }
 
 func init() {
-	gob.Register(&ArrayType{})
-	gob.Register(&ChanType{})
-	gob.Register(&FuncType{})
-	gob.Register(&MapType{})
-	gob.Register(&NamedType{})
-	gob.Register(&PointerType{})
+	gob.RegisterName(pkgPath+".ArrayType", &ArrayType{})
```

Comment:
Should the registered named be prefixed with a "*"? From my understanding gob.Register is only misbehaving (i.e. no fully qualifying) for named pointer types.

Reply:
From gob's implementation, the registered type will be firstly reflected for the actual type (userType), then stored in two map, with `<name, type>` and `<type, name>` pairs: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/encoding/gob/type.go;l=836-844;drc=f80e270babcddd28e29a52c1cb834ffed427bd8a

In the encode part, encoder will use the same `userType` again to reflect the actual type, and look up the name: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/encoding/gob/encode.go;l=400-405;drc=6a51c000def084b820756f066397fd2505e75e25

In the decode part, a similar but reverse process will be used for looking up type by name: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/encoding/gob/decode.go;l=699-704;drc=f62c9701b4bc61da6a5f4db8ef81d816f112430e

The default `gob.Register` call infers the type name with this: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/encoding/gob/type.go;l=854-856;drc=f80e270babcddd28e29a52c1cb834ffed427bd8a (without calling user type etc) . For instance, giving a pointer to `MapType` from two different packages, the name here will be inferred as the same value (yes it will have a `*` prefix as it's a pointer type). As a result, we will see panic in later register when using like issue #51  . 

The fix here is to add a qualified prefix to the value type, so we can avoid the panic on register. But to gob, it's just an unique identifier. Or, we can define it as something like `{pkgPath}.*MapType`. Let me know your thought!